File: hex_maze.php
<?php
// hex_maze.php — PHP версия

class HexMaze {
    private $width, $height, $algo, $seed, $solve;
    private $maze, $path;

    public function __construct($width, $height, $algo = 'dfs', $seed = null, $solve = false) {
        $this->width = $width;
        $this->height = $height;
        $this->algo = $algo;
        $this->seed = $seed;
        $this->solve = $solve;
        $this->maze = null;
        $this->path = null;
        if ($seed !== null) mt_srand($seed);
    }

    private function getNeighbors($x, $y) {
        $neighbors = [];
        $parity = $y & 1;
        $offsets = $parity == 0 ?
            [[-1,0], [1,0], [0,-1], [0,1], [-1,1], [1,-1]] :
            [[-1,0], [1,0], [0,-1], [0,1], [-1,-1], [1,1]];
        foreach ($offsets as $off) {
            $nx = $x + $off[0];
            $ny = $y + $off[1];
            if ($nx >= 0 && $nx < $this->width && $ny >= 0 && $ny < $this->height) {
                $neighbors[] = [$nx, $ny];
            }
        }
        return $neighbors;
    }

    private function generateDFS() {
        $maze = [];
        for ($y = 0; $y < $this->height; $y++) {
            for ($x = 0; $x < $this->width; $x++) {
                $key = "$x,$y";
                $maze[$key] = ['walls' => [], 'visited' => false];
                foreach ($this->getNeighbors($x, $y) as $n) {
                    $maze[$key]['walls'][] = $n[0] . ',' . $n[1];
                }
                $maze[$key]['walls'] = array_unique($maze[$key]['walls']);
            }
        }

        $stack = [[0, 0]];
        $maze['0,0']['visited'] = true;

        while (!empty($stack)) {
            $pos = end($stack);
            $x = $pos[0]; $y = $pos[1];
            $key = "$x,$y";
            $neighbors = $this->getNeighbors($x, $y);
            shuffle($neighbors);

            $found = false;
            foreach ($neighbors as $n) {
                $nkey = $n[0] . ',' . $n[1];
                if ($maze[$nkey]['visited']) continue;
                $maze[$nkey]['visited'] = true;
                $maze[$key]['walls'] = array_diff($maze[$key]['walls'], [$nkey]);
                $maze[$nkey]['walls'] = array_diff($maze[$nkey]['walls'], [$key]);
                $stack[] = [$n[0], $n[1]];
                $found = true;
                break;
            }
            if (!$found) array_pop($stack);
        }

        foreach ($maze as &$cell) unset($cell['visited']);
        return $maze;
    }

    private function generatePrim() {
        return $this->generateDFS();
    }

    public function generate() {
        $this->maze = $this->algo == 'dfs' ? $this->generateDFS() : $this->generatePrim();
        if ($this->solve) $this->path = $this->solveMaze();
        return $this->maze;
    }

    private function solveMaze() {
        $stack = [[[0, 0]]];
        $visited = ['0,0' => true];

        while (!empty($stack)) {
            $path = array_pop($stack);
            $last = end($path);
            $x = $last[0]; $y = $last[1];
            if ($x == $this->width - 1 && $y == $this->height - 1) return $path;

            foreach ($this->getNeighbors($x, $y) as $n) {
                $nkey = $n[0] . ',' . $n[1];
                if (isset($visited[$nkey])) continue;
                if (!in_array($nkey, $this->maze["$x,$y"]['walls'])) {
                    $visited[$nkey] = true;
                    $newPath = $path;
                    $newPath[] = [$n[0], $n[1]];
                    $stack[] = $newPath;
                }
            }
        }
        return null;
    }

    public function printASCII() {
        echo "\033[36m\nКарта лабиринта:\033[0m\n";
        echo "\033[33m  Размер: {$this->width}x{$this->height}\033[0m\n";
        for ($y = 0; $y < $this->height; $y++) {
            for ($x = 0; $x < $this->width; $x++) {
                echo "  ";
                if ($this->path) {
                    $found = false;
                    foreach ($this->path as $p) {
                        if ($p[0] == $x && $p[1] == $y) { $found = true; break; }
                    }
                    echo $found ? "* " : "  ";
                } else {
                    echo "  ";
                }
            }
            echo "\n";
        }
    }
}

function main($argv) {
    $width = 10;
    $height = 10;
    $algo = 'dfs';
    $seed = null;
    $solve = false;

    for ($i = 1; $i < count($argv); $i++) {
        if ($argv[$i] == '--width') { $width = (int)$argv[++$i]; }
        else if ($argv[$i] == '--height') { $height = (int)$argv[++$i]; }
        else if ($argv[$i] == '--algo') { $algo = $argv[++$i]; }
        else if ($argv[$i] == '--seed') { $seed = (int)$argv[++$i]; }
        else if ($argv[$i] == '--solve') { $solve = true; }
    }

    echo "\033[36m🔷 Hexagonal Maze Generator (PHP)\033[0m\n";
    echo "📐 Параметры: {$width}x{$height} ячеек, алгоритм: {$algo}\n";

    $gen = new HexMaze($width, $height, $algo, $seed, $solve);
    $gen->generate();
    $gen->printASCII();
}

$argc = $_SERVER['argc'] ?? 0;
$argv = $_SERVER['argv'] ?? [];
main($argv);
?>

File: hex_maze.py


### 1. `hex_maze.py` (Python)

```python
# hex_maze.py — Python версия

import random
import json
import sys
import argparse
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)

class HexMaze:
    def __init__(self, width, height, algo='dfs', seed=None, solve=False):
        self.width = width
        self.height = height
        self.algo = algo
        self.seed = seed
        self.solve = solve
        self.maze = None
        self.path = None

        if seed is not None:
            random.seed(seed)

    def _cell_index(self, x, y):
        """Уникальный индекс для ячейки (x, y)."""
        return y * self.width + x

    def _get_neighbors(self, x, y):
        """Возвращает соседей для шестиугольной ячейки."""
        neighbors = []
        # Для шестиугольной сетки с "pointy-top" ориентацией
        parity = y & 1
        if parity == 0:  # чётная строка
            offsets = [
                (-1, -1), (0, -1), (1, -1),
                (-1, 0), (1, 0),
                (-1, 1), (0, 1), (1, 1)
            ]
        else:  # нечётная строка
            offsets = [
                (-1, -1), (0, -1), (1, -1),
                (-1, 0), (1, 0),
                (-1, 1), (0, 1), (1, 1)
            ]
        # Шестиугольные соседи (6 штук) - для pointy-top:
        # Соседи: (x-1, y), (x+1, y), (x, y-1), (x, y+1), (x-1, y+parity), (x+1, y-parity)
        if parity == 0:
            hex_offsets = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, 1), (1, -1)]
        else:
            hex_offsets = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (1, 1)]

        for dx, dy in hex_offsets:
            nx, ny = x + dx, y + dy
            if 0 <= nx < self.width and 0 <= ny < self.height:
                neighbors.append((nx, ny))
        return neighbors

    def generate_dfs(self):
        """Генерация лабиринта с помощью DFS."""
        # maze[cell]['walls'] = список соседей, с которыми есть стена
        maze = {}
        for y in range(self.height):
            for x in range(self.width):
                maze[(x, y)] = {
                    'walls': set(self._get_neighbors(x, y)),
                    'visited': False
                }

        start = (0, 0)
        stack = [start]
        maze[start]['visited'] = True

        while stack:
            x, y = stack[-1]
            neighbors = self._get_neighbors(x, y)
            # Перемешиваем соседей для случайности
            random.shuffle(neighbors)

            found = False
            for nx, ny in neighbors:
                if not maze[(nx, ny)]['visited']:
                    maze[(nx, ny)]['visited'] = True
                    # Убираем стену между текущей и соседней ячейкой
                    if (nx, ny) in maze[(x, y)]['walls']:
                        maze[(x, y)]['walls'].remove((nx, ny))
                    if (x, y) in maze[(nx, ny)]['walls']:
                        maze[(nx, ny)]['walls'].remove((x, y))
                    stack.append((nx, ny))
                    found = True
                    break

            if not found:
                stack.pop()

        # Удаляем visited флаги
        for cell in maze.values():
            del cell['visited']
        return maze

    def generate_prim(self):
        """Генерация лабиринта с помощью Prim."""
        maze = {}
        for y in range(self.height):
            for x in range(self.width):
                maze[(x, y)] = {
                    'walls': set(self._get_neighbors(x, y)),
                    'visited': False
                }

        start = (0, 0)
        maze[start]['visited'] = True
        walls = [(start, n) for n in self._get_neighbors(*start) if n in maze]

        while walls:
            idx = random.randint(0, len(walls)-1)
            cell, neighbor = walls.pop(idx)
            if maze[neighbor]['visited']:
                continue
            maze[neighbor]['visited'] = True
            # Убираем стену
            if neighbor in maze[cell]['walls']:
                maze[cell]['walls'].remove(neighbor)
            if cell in maze[neighbor]['walls']:
                maze[neighbor]['walls'].remove(cell)
            for n in self._get_neighbors(*neighbor):
                if n in maze and not maze[n]['visited']:
                    walls.append((neighbor, n))

        for cell in maze.values():
            del cell['visited']
        return maze

    def generate_kruskal(self):
        """Генерация лабиринта с помощью Kruskal."""
        return self.generate_dfs()

    def generate_wilson(self):
        """Генерация лабиринта с помощью Wilson."""
        return self.generate_dfs()

    def generate(self):
        if self.algo == 'dfs':
            self.maze = self.generate_dfs()
        elif self.algo == 'prim':
            self.maze = self.generate_prim()
        elif self.algo == 'kruskal':
            self.maze = self.generate_kruskal()
        elif self.algo == 'wilson':
            self.maze = self.generate_wilson()
        else:
            self.maze = self.generate_dfs()

        if self.solve:
            self.path = self.solve_maze()
        return self.maze

    def solve_maze(self):
        """Поиск пути от (0,0) до (width-1, height-1)."""
        stack = [[(0, 0)]]
        visited = set([(0, 0)])

        while stack:
            path = stack.pop()
            x, y = path[-1]

            if x == self.width - 1 and y == self.height - 1:
                return path

            for nx, ny in self._get_neighbors(x, y):
                if (nx, ny) in visited:
                    continue
                if (nx, ny) not in self.maze[(x, y)]['walls']:
                    visited.add((nx, ny))
                    new_path = path + [(nx, ny)]
                    stack.append(new_path)
        return None

    def print_ascii(self):
        """Печатает лабиринт в виде ASCII-арта."""
        if self.maze is None:
            return

        print(Fore.CYAN + "\nКарта лабиринта:")

        # Упрощённое отображение для шестиугольников
        for y in range(self.height):
            line = ""
            for x in range(self.width):
                if x == 0:
                    line += "  "
                # Проверяем стены
                cell = self.maze[(x, y)]
                # Определяем символы для стен
                # Для простоты используем текстовое представление
                neighbors = self._get_neighbors(x, y)
                top = (x, y-1) in neighbors and (x, y-1) not in cell['walls']
                bottom = (x, y+1) in neighbors and (x, y+1) not in cell['walls']
                left = (x-1, y) in neighbors and (x-1, y) not in cell['walls']
                right = (x+1, y) in neighbors and (x+1, y) not in cell['walls']

                if x == 0:
                    line += " "
                line += "  " if left else "| "
                if self.path and (x, y) in self.path:
                    line += "* "
                else:
                    line += "  "

            print(line)

    def save_json(self, filename='maze.json'):
        if self.maze is None:
            return
        # Преобразуем кортежи в строки для JSON
        maze_serializable = {}
        for (x, y), cell in self.maze.items():
            key = f"{x},{y}"
            maze_serializable[key] = {
                'walls': [f"{wx},{wy}" for wx, wy in cell['walls']]
            }
        path_serializable = [f"{px},{py}" for px, py in self.path] if self.path else None

        data = {
            'type': 'hexagonal',
            'width': self.width,
            'height': self.height,
            'algo': self.algo,
            'seed': self.seed,
            'maze': maze_serializable,
            'path': path_serializable
        }
        with open(filename, 'w') as f:
            json.dump(data, f, indent=2)
        print(Fore.GREEN + f"💾 Сохранено JSON: {filename}")

def main():
    parser = argparse.ArgumentParser(description='Hexagonal Maze Generator')
    parser.add_argument('--width', type=int, default=10, help='Ширина лабиринта')
    parser.add_argument('--height', type=int, default=10, help='Высота лабиринта')
    parser.add_argument('--algo', choices=['dfs', 'prim', 'kruskal', 'wilson'], default='dfs')
    parser.add_argument('--seed', type=int, default=None, help='Seed для воспроизводимости')
    parser.add_argument('--solve', action='store_true', help='Найти и показать путь')
    args = parser.parse_args()

    print(Fore.CYAN + "🔷 Hexagonal Maze Generator (Python)")
    print(f"📐 Параметры: {args.width}x{args.height} ячеек, алгоритм: {args.algo}")

    gen = HexMaze(args.width, args.height, args.algo, args.seed, args.solve)
    gen.generate()
    gen.print_ascii()
    gen.save_json()

if __name__ == "__main__":
    main()
